database.cc

```diff
@@ -2321,6 +2321,71 @@ future<> database::drain() {
+                        // "snapshots" directory
+
+                        return lister::scan_dir(parent_dir / de.name, *dirs_only_entries_ptr, [this, &details, ks_name, cf_name, cf_dir] (fs::path parent_dir, directory_entry de) {
+                            sstring snapshot_name = de.name;
+                            std::pair<sstring, sstring> cf_map_key = { ks_name, cf_name };
+                            std::unordered_map<std::pair<sstring, sstring>, snapshot_ctl::snapshot_details, utils::tuple_hash> cf_map = {{ cf_map_key, {0, 0, cf_name, ks_name} }};
+                            details.emplace(snapshot_name, std::move(cf_map));
+
+                            return lister::scan_dir(parent_dir / de.name,  { directory_entry_type::regular }, [this, &details, ks_name, cf_name, cf_map_key, cf_dir, snapshot_name = std::move(snapshot_name)] (fs::path snapshot_dir, directory_entry de) {
+                                return io_check(file_size, (snapshot_dir / de.name).native()).then([this, cf_dir, cf_map_key, &details, snapshot_name, snapshot_dir, name = de.name] (auto size) {
+                                    // The manifest is the only file expected to be in this directory not belonging to the SSTable.
+                                    // For it, we account the total size, but zero it for the true size calculation.
+                                    //
+                                    // All the others should just generate an exception: there is something wrong, so don't blindly
+                                    // add it to the size.
+                                    if (name != "manifest.json" && name != "schema.cql") {
+                                        sstables::entry_descriptor::make_descriptor(snapshot_dir.native(), name);
+                                        details.at(snapshot_name).at(cf_map_key).total += size;
+                                    } else {
+                                        size = 0;
+                                    }
+                                    return io_check(file_size, (fs::path(cf_dir) / name).native()).then_wrapped([cf_map_key, &details, snapshot_name, size] (auto fut) {
+                                        try {
+                                            // File exists in the main SSTable directory. Snapshots are not contributing to size
+                                            fut.get0();
+                                        } catch (std::system_error& e) {
+                                            if (e.code() != std::error_code(ENOENT, std::system_category())) {
+                                                throw;
+                                            }
+                                            details.at(snapshot_name).at(cf_map_key).live += size;
+                                        }
```

Comment:
Is there a way to check that a file is not present in the sstable directory without throwing exceptions that are slow?

Reply:
On a second thought this is probably alright and we can do improvements as a follow up if needed.